### PR TITLE
feat(typeorm): composite primary keys via @entity(pk: [...])

### PR DESCRIPTION
**Pull request by abernatskiy**
Entities can only be keyed on `id`. This adds an opt-in composite primary key over `id` plus additional non-nullable scalar or enum columns.

The motivating case is tables partitioned on a time column: the partitioning column has to participate in every unique constraint, so `id` alone cannot be the key. Today that forces a fork of `typeorm-codegen`, since `id` is hardcoded as the sole primary column.

```graphql
type Transfer @entity(pk: ["id", "timestamp"]) {
  id: ID!
  timestamp: DateTime!
  amount: BigInt!
}
```

```ts
@PrimaryColumn_()                                  id!: string
@DateTimeColumn_({primary: true, nullable: false}) timestamp!: Date
@BigIntColumn_({nullable: false})                  amount!: bigint
```

## What changed

**openreader**
- `pk` argument on `@entity`, validated at the GraphQL layer and again in `validateModel` for programmatically built models.
- `id` must lead the key, so the leading column of the primary key index is unchanged and id-prefix lookups keep their plan.
- `JSON` and `Bytes` are rejected as key columns — neither survives the round trip through the hot-block change log as a comparable value.
- New `<entity>ByPks` query taking one argument per key column. `<entity>ById` is unchanged and still supported for single-column keys. Composite-key entities expose `ByPks` only, because `id` alone does not identify a row there.
- `queryName` continues to rename the single-entity lookup — `ById` for a single-column key, `ByPks` for a composite one — with a guard against the one case where the override would collide with the generated name.
- The `thegraph` dialect keeps its subgraph-style query name and takes the extra key columns as arguments.

**typeorm-codegen**
- `id` keeps `@PrimaryColumn_()`; the remaining key columns are marked `{primary: true}` on the typed decorator they already had, so type mapping, transformers and index annotations are untouched.
- Output for entities without `pk` is byte-identical — asserted by a test that generates the same entity with and without `pk: ["id"]` and compares the files.

**typeorm-store**
- `primary` option on the typed column decorators. `PrimaryColumn`'s `id`-only assertion is retained; composite columns do not route through it.
- Upserts take their `ON CONFLICT` target from the entity's key rather than assuming `id`.
- Hot-block change records carry the full key, so a rollback addresses one row instead of every row sharing an `id`. Records for single-column keys serialize exactly as before, so a hot window written by an earlier version stays readable across an upgrade.

## Compatibility

Additive. Entities that do not declare `pk` generate identical models, identical SQL and an identical GraphQL surface apart from the added `ByPks` query.

## Testing

- openreader: 186 passing, on PostgreSQL and CockroachDB.
- typeorm-store: 76 passing, on PostgreSQL and CockroachDB.
- typeorm-codegen: 54 passing.

New coverage includes rollback of inserts, updates and deletes under a composite key; a three-column key whose last column carries a value transformer; insert-vs-update discrimination when rows share an `id`; and the query-name rules above. The rollback and transformer tests were checked against a mutated build to confirm they fail without the corresponding change.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01Axzv4syC31YvaTQMGfuJJZ

### CODE REVIEW

File: typeorm/typeorm-store/src/hot.ts [hunk 1]
```diff
@@ -306,6 +324,87 @@ export async function rollbackBlock(
...
+ */
+function keyPredicate(em: EntityManager, ch: ChangeRecord, firstParam: number): {sql: string; values: any[]} {
+    let sql = `id = $${firstParam}`
+    let values: any[] = [ch.id]
+    for (let column in ch.key) {
+        values.push(ch.key[column])
+        sql += ` AND ${escapeIdentifier(em, column)} = $${firstParam + values.length - 1}`
+    }
+    return {sql, values}
+}
```
[the remark this answers is not in the document]
Comment (abernatskiy) [reply typeorm/typeorm-store/src/hot.ts:346]: **(With Claude):** I don't think this one holds — `for...in` over `undefined` is a spec-defined no-op rather than a `TypeError`.

Per ECMA-262 `ForIn/OfHeadEvaluation`: if the expression evaluates to `undefined` or `null`, the loop completes immediately with zero iterations. Empirically:

```js
> let ch = {id: 'a'}; let n = 0; for (const c in ch.key) n++; n
0
```

`Object.keys(undefined)` *does* throw, which may be what the finding has in mind — but `keyPredicate()` doesn't call it. With `ch.key` absent the function returns exactly the previous predicate, `id = $1` with `[ch.id]`, which is why single-column-key records were left unchanged on the wire.

This path is also well covered rather than untested: every pre-existing hot-block rollback test drives single-column-key entities through `keyPredicate()` with `ch.key === undefined` — `data-type-rollback`, `fork-cascade`, `fork-deep`, `hot-edge-cases`, `mixed-boundary`, `crash-recovery` and `transact-parity` among them. The full `typeorm-store` suite (76 tests) passes on both PostgreSQL and CockroachDB, and a rollback regression here would take most of it down.

Resolving as a false positive — happy to reopen if I've misread the concern.
